### PR TITLE
[Improvement](statistics) Improve sample count accuracy.

### PR DESCRIPTION
**Pull request by Jibing-Li**
While doing sample analyze, the result of row count, null number and datasize need to multiply a coefficient based on the sample percent/rows. This pr is mainly to calculate the coefficient according to the sampled file size over total size.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/catalog/TableIf.java [hunk 1]
```diff
@@ -142,6 +142,8 @@ default int getBaseColumnIdxByName(String colName) {
 
     Map<String, Set<String>> findReAnalyzeNeededPartitions();
 
+    List<Long> getChunkSizes();
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/TableIf.java:145]: Add comment for these new interface

File: fe/fe-core/src/main/java/org/apache/doris/catalog/external/ExternalTable.java [hunk 2]
```diff
@@ -396,4 +396,9 @@ public Map<String, Set<String>> findReAnalyzeNeededPartitions() {
         partitions.add("Dummy Partition");
         return getBaseSchema().stream().collect(Collectors.toMap(Column::getName, k -> partitions));
     }
+
+    @Override
+    public List<Long> getChunkSizes() {
+        return null;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/external/ExternalTable.java:402]: Better not return "null"

File: fe/fe-core/src/main/java/org/apache/doris/statistics/HMSAnalysisTask.java [hunk 3]
```diff
@@ -58,14 +58,15 @@ public class HMSAnalysisTask extends BaseAnalysisTask {
             + "${idxId} AS idx_id, "
             + "'${colId}' AS col_id, "
             + "NULL AS part_id, "
-            + "${countExpr} AS row_count, "
-            + "NDV(`${colName}`) AS ndv, "
-            + "${nullCountExpr} AS null_count, "
+            + "ROUND(COUNT(1) * ${scaleFactor}) AS row_count, "
+            + "case when NDV(`${colName}`)/count('${colName}') < 0.3 then NDV(`${colName}`) "
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/HMSAnalysisTask.java:62]: this factor should be defined some where

File: fe/fe-core/src/main/java/org/apache/doris/statistics/HMSAnalysisTask.java [hunk 4]
```diff
@@ -277,38 +277,53 @@ private Map<String, String> buildTableStatsParams(String partId) {
...
-            return "SUM(CASE WHEN `${colName}` IS NULL THEN 1 ELSE 0 END)";
+            return String.format("TABLESAMPLE(%d ROWS)", tableSample.getSampleValue());
         }
     }
 
-    protected String getDataSizeFunction(Column column) {
-        String originFunction = super.getDataSizeFunction(column);
-        if (info.samplePercent > 0 && !isPartitionOnly) {
-            return String.format("ROUND((%s) * 100 / %d)", originFunction, info.samplePercent);
+    protected String getSampleScaleFactor() {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/HMSAnalysisTask.java:300]: Add some comment to explain the logic in code, better give some example

File: fe/fe-core/src/main/java/org/apache/doris/catalog/external/ExternalTable.java (hunk 2 above)
Comment (Jibing-Li) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/external/ExternalTable.java:402]: throw NotImplementedException instead

**Comment by Jibing-Li** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/statistics/util/StatisticsUtil.java [hunk 5]
```diff
@@ -574,31 +574,71 @@ public static long getRowCountFromFileList(HMSExternalTable table) {
...
             // because it has enough space to keep partition info of all tables in cache.
             partitionValues = cache.getPartitionValues(table.getDbName(), table.getName(), partitionColumnTypes);
         }
+        return partitionValues;
+    }
+
+    public static List<HiveMetaStoreCache.FileCacheValue> getFilesForPartitions(
+            HMSExternalTable table, HiveMetaStoreCache.HivePartitionValues partitionValues, int sampleSize) {
+        if (table.isView()) {
+            return null;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/util/StatisticsUtil.java:626]: Better not return null

File: fe/fe-core/src/main/java/org/apache/doris/catalog/external/HMSExternalTable.java [hunk 6]
```diff
@@ -635,6 +636,30 @@ public void gsonPostProcess() throws IOException {
...
+            for (HiveMetaStoreCache.HiveFileStatus file : files.getFiles()) {
+                result.add(file.getLength());
+            }
+        }
+        return result;
+    }
+
+    @Override
+    public long getDataSize(boolean singleReplica) {
+        List<Long> chunkSizes = getChunkSizes();
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/external/HMSExternalTable.java:656]: Looks like it is a very heavy operation? It need to fetch all partitions and all files of a table, just for getting the total size of this table?
But in most case, the total size of table can be found in hive table's property?

Comment (Jibing-Li) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/external/HMSExternalTable.java:656]: It is a heavy operation, as we discussed earlier, this brings a redundant fetching of all the files.
Usually  we can get total size of hive table in hms, but this call is not only to get the total size, but the size of each file. We use it to calculate the accurate sample ratio.

Comment (Jibing-Li) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/external/HMSExternalTable.java:671]: Modified it to get total size from HMS first, only to fetch all files' size when HMS doesn't contain the total size value.

**Comment by Jibing-Li** [comment]
run buildall

**Comment by Jibing-Li** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Review by Kikyou1997 — APPROVED** [review]

**Pull request merged**